### PR TITLE
Improve Serve API Input Validations

### PR DESCRIPTION
**Pull request by mfitton**
## Why are these changes needed?

This PR improves error messages when a user makes a mistake by specifying a backend or endpoint name that does not exist. It also adds unit tests to validate that several actions are not allowed:
1. You cannot create two endpoints with the same name
2. You cannot create two endpoints with the same route
3. You cannot create two backends with the same name.

## Related issue number
No related issue.

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/latest/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested (please justify below)

### CODE REVIEW

**Comment by mfitton** [comment]
@simon-mo this is my first PR! Please give me any advice you might have.

File: python/ray/serve/master.py [hunk 1]
```diff
@@ -219,20 +219,30 @@ def get_all_endpoints(self):
...
-            assert (backend in self.backend_table.list_backends()
-                    ), "backend {} is not registered".format(backend)
-        assert np.isclose(
-            prob, 1, atol=0.02
-        ), "weights must sum to 1, currently it sums to {}".format(prob)
+            assert backend in existing_backends, \
+                f"Attempted to assign traffic to a backend '{backend}' that " \
+                "is not registered"
+
+            assert np.isclose(
```
Comment (simon-mo) [inline python/ray/serve/master.py:243]: Should this be de-dented? Prob should only sum up to one after we add up all the backends right?

File: python/ray/serve/master.py [hunk 2]
```diff
@@ -255,6 +270,10 @@ async def create_backend(self, backend_tag, backend_config, func_or_class,
         backend_config_dict = dict(backend_config)
         backend_worker = create_backend_worker(func_or_class)
 
+        assert backend_tag not in self.get_all_backends(), \
+            f"Cannot create backend '{backend_tag}' because a backend" \
+            "with that name already exists."
+
```
Comment (simon-mo) [inline python/ray/serve/master.py:276]: f-string -> str.format

**Review by simon-mo — COMMENTED** [review]
Nice!

File: python/ray/serve/master.py (hunk 1 above)
Comment (mfitton) [reply python/ray/serve/master.py:243]: Yes, it should, thanks. I didn't mean to indent it, and I've undone that.

File: python/ray/serve/master.py [hunk 3]
```diff
@@ -219,20 +219,30 @@ def get_all_endpoints(self):
...
+        existing_backends = set(self.get_all_backends())
         for backend, weight in traffic_policy_dictionary.items():
             prob += weight
-            assert (backend in self.backend_table.list_backends()
-                    ), "backend {} is not registered".format(backend)
-        assert np.isclose(
-            prob, 1, atol=0.02
-        ), "weights must sum to 1, currently it sums to {}".format(prob)
+            assert backend in existing_backends, \
+                f"Attempted to assign traffic to a backend '{backend}' that " \
```
[the remark this answers is not in the document]
Comment (mfitton) [reply python/ray/serve/master.py]: done

File: python/ray/serve/master.py (hunk 2 above)
Comment (mfitton) [reply python/ray/serve/master.py]: done

File: python/ray/serve/master.py [hunk 4]
```diff
@@ -242,6 +252,11 @@ async def set_traffic(self, endpoint_name, traffic_policy_dictionary):
                                         traffic_policy_dictionary)
 
     async def create_endpoint(self, route, endpoint_name, methods):
+        err_prefix = "Cannot create endpoint. "
+        assert route not in self.get_all_routes(), \
+            f"{err_prefix} Route '{route}' is already registered."
+        assert endpoint_name not in self.get_all_endpoints(), \
+            f"{err_prefix} Endpoint '{endpoint_name}' is already registered."
```
[the remark this answers is not in the document]
Comment (mfitton) [reply python/ray/serve/master.py]: done

**Review by simon-mo — APPROVED** [review]

**Pull request merged**